### PR TITLE
cache open graph response

### LINKED ISSUE #17005: Avoid unnecessary network calls when fetching OpenGraph metadata when editing a post
### Description
OpenGraph metadata is fetched from the supplied url in the `{% embed url %}` tag as a fallback if the provided URL doesn't exist as a "supported" embed tag. The network request happens when saving and/or previewing a post. Ideally, we'd want to avoid making a lot of unnecessary network calls to the supplied URL while an author is drafting and/or editing their post.

### Why is this an issue?
This is to improve the UX as an author. If kept as is, the experience could be less than ideal as the author waits until the server fetches the OpenGraph data every time a preview/save happens. This is especially true when editing/drafting a post with several embedded URLs that fall back to OpenGraph.

### Possible Solution
When a post is saved, we can temporarily cache the response for a brief time period. We'd then check the cache first to see if there is any data present and use that to render the preview. The solution should support multiple URLs in a single post.

### PR DESCRIPTION
**Pull request by dscottS3**
## What type of PR is this? (check all applicable)

- [x] Optimization

## Description

This PR enhances the UX for authors by decreasing load time by caching the HTML response when an author embeds a url that uses OpenGraph. The OpenGraph data is then parsed from the cached HTML. Current cache expiration is set to 15 minutes.

**Note:** This was put together pretty quick and probably only a small percentage of folks will benefit initially. That said, the thought here is to build upon this as things grow and change.

## Related Tickets & Documents

- Closes #17005 

## QA Instructions, Screenshots, Recordings

- Draft a post and embed a "non-supported" url such as `https://weather.com`
- Open Inspector and view network tab (may have to check "preserve log")
- Click "Preview" tab
- Take note of the time it takes to load
- Flip back to Edit, then click Preview again
- Check the time it takes to load; should be a lot faster

### UI accessibility concerns?

No

## Added/updated tests?

- [x] No, and this is why: This is a private method and it's a core piece of functionality (`Rails.cache.fetch`) that's being used everywhere. I'm not sure what a test here would gain us considering it's just writing/fetching a cache key. I can be convinced otherwise.

## [Forem core team only] How will this change be communicated?

- [x] I will share this change internally with the appropriate teams

### CODE REVIEW

File: app/services/open_graph.rb [hunk 1]
```diff
@@ -9,9 +9,13 @@ class OpenGraph
...
 
   DEFAULT_METHODS = %i[description title url].freeze
+  CACHE_EXPIRY_IN_MINUTES = 15
 
   def initialize(url)
-    @page = MetaInspector.new(url)
+    # get the html from cache or fetch new
+    html = fetch_html(url)
+    # MetaInspector can read html
+    @page = MetaInspector.new(url, document: html)
```
Comment (Ridhwana) [inline app/services/open_graph.rb:18]: I'm wondering if the two comments are necessary here, as you've done a great job naming the method and instance variable.

**Review by Ridhwana — APPROVED** [review]
Looks good, just left a small comment about the comments 😅

File: app/services/open_graph.rb (hunk 1 above)
Comment (jeremyf) [reply app/services/open_graph.rb:18]: I'm with Ridhwana on this.  The comments echo what the code does, you could ✂️ these inline comments unless there's some deeper "Why am I doing this" you want to convey.

**Comment by jeremyf** [comment]
> No, and this is why: This is a private method and it's a core piece of functionality (Rails.cache.fetch) that's being used everywhere. I'm not sure what a test here would gain us considering it's just writing/fetching a cache key. I can be convinced otherwise.

Love the explanation and affirm the decision to eschew a test.

**Review by jeremyf — APPROVED** [review]

**Pull request merged**